File: GeoBroker-Commons/src/main/java/de/hasenburg/geobroker/commons/model/spatial/Geofence.java
package de.hasenburg.geobroker.commons.model.spatial;

import de.hasenburg.geobroker.commons.exceptions.RuntimeShapeException;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.locationtech.spatial4j.io.ShapeWriter;
import org.locationtech.spatial4j.shape.*;

import java.text.ParseException;
import java.util.List;
import java.util.Objects;

import static de.hasenburg.geobroker.commons.model.spatial.SpatialContext.GEO;


public class Geofence{

	private static final Logger logger = LogManager.getLogger();
	private final Shape shape;

	// TODO increase size a little bit so that we do not miss any due to rounding issues
	// we need it most times anyways, so let's buffer it //
	public final Rectangle boundingBox;

	private Geofence(Shape shape) {
		this.shape = shape;
		this.boundingBox = shape.getBoundingBox();
	}

	public Geofence(String wkt) throws ParseException {
		WKTReader reader = (WKTReader) GEO.getFormats().getWktReader();
		this.shape = reader.parse(wkt);
		this.boundingBox = this.shape.getBoundingBox();
	}

	/**
	 * Creates a new geofence based on the supplied locations
	 *
	 * @param surroundingLocations - the locations that surround the geofence
	 * @return a new geofence
	 * @throws RuntimeShapeException if invalid param
	 */
	public static Geofence polygon(List<Location> surroundingLocations) {
		if (surroundingLocations.size() < 3) {
			throw new RuntimeShapeException("A geofence needs at least 3 locations");
		}

		final ShapeFactory.PolygonBuilder polygonBuilder = GEO.getShapeFactory().polygon();
		for (Location location : surroundingLocations) {
			polygonBuilder.pointLatLon(location.getLat(), location.getLon());
		}
		// close polygon
		polygonBuilder.pointLatLon(surroundingLocations.get(0).getLat(), surroundingLocations.get(0).getLon());
		return new Geofence(polygonBuilder.build());
	}

	public static Geofence rectangle(Location southWest, Location northEast) {
		Rectangle r = GEO.getShapeFactory().rect(southWest.getPoint(), northEast.getPoint());
		return new Geofence(r);
	}

	public static Geofence circle(Location location, double radiusDegree) {
		Circle c = GEO.getShapeFactory().circle(location.getPoint(), radiusDegree);
		return new Geofence(c);
	}

	public static Geofence world() {
		Shape worldShape = GEO.getWorldBounds();
		return new Geofence(worldShape);
	}

	public boolean isRectangle() {
		return GEO.getShapeFactory().getGeometryFrom(shape).isRectangle();
	}

	/*****************************************************************
	 * BoundingBox
	 ****************************************************************/

	public Location getBoundingBoxNorthWest() {
		return new Location(boundingBox.getMaxY(), boundingBox.getMinX());
	}

	public Location getBoundingBoxNorthEast() {
		return new Location(boundingBox.getMaxY(), boundingBox.getMaxX());
	}

	public Location getBoundingBoxSouthEast() {
		return new Location(boundingBox.getMinY(), boundingBox.getMaxX());
	}

	public Location getBoundingBoxSouthWest() {
		return new Location(boundingBox.getMinY(), boundingBox.getMinX());
	}

	/**
	 * See {@link Rectangle#getHeight()}, is the latitude distance in degree
	 */
	public double getBoundingBoxLatDistanceInDegree() {
		return boundingBox.getHeight();
	}

	/**
	 * See {@link Rectangle#getWidth()}, is the longitude distance in degree
	 */
	public double getBoundingBoxLonDistanceInDegree() {
		return boundingBox.getWidth();
	}

	public boolean contains(Location location) {
		if (location.isUndefined()) {
			return false;
		}
		return shape.relate(location.getPoint()).equals(SpatialRelation.CONTAINS);
	}

	/**
	 * For us, intersects is an "intersection" but also something more specific such as "contains" or within.
	 */
	public boolean intersects(Geofence geofence) {
		SpatialRelation sr = shape.relate(geofence.shape);
		return sr.equals(SpatialRelation.INTERSECTS) || sr.equals(SpatialRelation.CONTAINS) ||
				sr.equals(SpatialRelation.WITHIN);
	}

	public boolean disjoint(Geofence geofence) {
		return shape.relate(geofence.shape).equals(SpatialRelation.DISJOINT);
	}

	/*****************************************************************
	 * Getters and String
	 ****************************************************************/

	public Shape getShapeObject() {
		return this.shape;
	}

	public String getWKTString() {
		ShapeWriter writer = GEO.getFormats().getWktWriter();
		return writer.toString(shape);
	}

	@Override
	public String toString() {
		return getWKTString();
	}

	/*****************************************************************
	 * Generated methods
	 ****************************************************************/

	@Override
	public boolean equals(Object o) {
		if (this == o) {
			return true;
		}
		if (o == null || getClass() != o.getClass()) {
			return false;
		}
		Geofence geofence = (Geofence) o;
		return Objects.equals(shape, geofence.shape);
	}

	@Override
	public int hashCode() {

		return Objects.hash(shape);
	}

	public static void main(String[] args) {
		Location paris = new Location(48.86, 2.35);
		Location berlin = new Location(52.52, 13.40);
		Geofence parisArea = Geofence.circle(paris, 3.0);
		Geofence berlinArea = Geofence.circle(berlin, 3.0);

		logger.info("Paris area = {}", parisArea);
		logger.info("Berlin area = {}", berlinArea);
		logger.info("The areas intersect: {}", berlinArea.intersects(parisArea));

		Location justIn = new Location(45.87, 2.3);
		logger.info(parisArea.contains(justIn));

		logger.info("Contains check Benchmark");
		Geofence world = Geofence.world();
		long amount = 10000000;

		long time = System.nanoTime();
		for (int i = 0; i < amount; i++) {
			berlinArea.contains(berlin);
		}
		logger.info("{} berlin in circle checks per ms", amount * 1000 * 1000 / (System.nanoTime() - time));

		time = System.nanoTime();
		for (int i = 0; i < amount; i++) {
			berlinArea.contains(paris);
		}
		logger.info("{} berlin out circle checks per ms", amount * 1000 * 1000 / (System.nanoTime() - time));

		time = System.nanoTime();
		for (int i = 0; i < amount; i++) {
			world.contains(berlin);
		}
		logger.info("{} world checks per ms", amount * 1000 * 1000 / (System.nanoTime() - time));
	}
}

File: GeoBroker-Commons/src/main/kotlin/de/hasenburg/geobroker/commons/model/spatial/Geofence.kt
package de.hasenburg.geobroker.commons.model.spatial

import de.hasenburg.geobroker.commons.exceptions.RuntimeShapeException
import de.hasenburg.geobroker.commons.model.spatial.SpatialContextK.GEO
import kotlinx.serialization.*
import kotlinx.serialization.internal.StringDescriptor
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonConfiguration
import org.apache.logging.log4j.LogManager
import org.locationtech.spatial4j.shape.Rectangle
import org.locationtech.spatial4j.shape.Shape
import org.locationtech.spatial4j.shape.SpatialRelation

private val logger = LogManager.getLogger()

@Serializable
class Geofence(@Serializable(with = ShapeWKTSerializer::class) @SerialName("wkt") private val shape: Shape) {

    val center: Location
        get() = Location(shape.center)

    fun isRectangle(): Boolean {
        return GEO.shapeFactory.getGeometryFrom(shape).isRectangle
    }

    /*****************************************************************
     * Relationships
     ****************************************************************/

    /**
     * Returns true if this geofence contains the given [location].
     * Returns false if not or [location] == null.
     */
    fun contains(location: Location?): Boolean {
        return if (location != null) {
            shape.relate(location.point) == SpatialRelation.CONTAINS
        } else {
            false
        }
    }

    /**
     * For us, intersects is an "intersection" but also something more specific such as "contains" or within.
     */
    fun intersects(geofence: Geofence): Boolean {
        val sr = shape.relate(geofence.shape)
        return sr == SpatialRelation.INTERSECTS || sr == SpatialRelation.CONTAINS || sr == SpatialRelation.WITHIN
    }

    fun disjoint(geofence: Geofence): Boolean {
        return shape.relate(geofence.shape) == SpatialRelation.DISJOINT
    }

    /*****************************************************************
     * Bounding Box
     ****************************************************************/

    val boundingBox: Rectangle
        get() = shape.boundingBox

    val boundingBoxNorthWest: Location
        get() = Location(boundingBox.maxY, boundingBox.minX)
    val boundingBoxNorthEast: Location
        get() = Location(boundingBox.maxY, boundingBox.maxX)
    val boundingBoxSouthEast: Location
        get() = Location(boundingBox.minY, boundingBox.maxX)
    val boundingBoxSouthWest: Location
        get() = Location(boundingBox.minY, boundingBox.minX)

    /**
     * See [Rectangle.getHeight], is the latitude distance in degree
     */
    val boundingBoxLatDistanceInDegree: Double
        get() = boundingBox.height

    /**
     * See [Rectangle.getWidth], is the longitude distance in degree
     */
    val boundingBoxLonDistanceInDegree: Double
        get() = boundingBox.width

    /*****************************************************************
     * Others
     ****************************************************************/

    companion object {

        fun circle(location: Location, radiusDegree: Double): Geofence {
            return Geofence(GEO.shapeFactory.circle(location.point, radiusDegree))
        }

        fun rectangle(southWest: Location, northEast: Location): Geofence {
            return Geofence(GEO.shapeFactory.rect(southWest.point, northEast.point))
        }

        fun world(): Geofence {
            return Geofence(GEO.worldBounds)
        }

        fun fromWkt(wkt: String): Geofence {
            val reader = GEO.formats.wktReader as CustomWKTReader
            return Geofence(reader.parse(wkt) as Shape)
        }

        /**
         * Creates a new geofence based on the supplied locations
         *
         * @param surroundingLocations - the locations that surround the geofence
         * @return a new geofence
         * @throws RuntimeShapeException if less than three [surroundingLocations]
         */
        fun polygon(surroundingLocations: List<Location>): Geofence {
            if (surroundingLocations.size < 3) {
                throw RuntimeShapeException("A geofence needs at least 3 locations")
            }
            val polygonBuilder = GEO.shapeFactory.polygon()
            for (location in surroundingLocations) {
                polygonBuilder.pointLatLon(location.lat, location.lon)
            }
            // close polygon
            polygonBuilder.pointLatLon(surroundingLocations[0].lat, surroundingLocations[0].lon)
            return Geofence(polygonBuilder.build())
        }
    }

    override fun toString(): String {
        return GEO.formats.wktWriter.toString(shape)
    }

    /*****************************************************************
     * Generated methods
     ****************************************************************/

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as Geofence

        if (shape != other.shape) return false

        return true
    }

    override fun hashCode(): Int {
        return shape.hashCode()
    }

}

/*****************************************************************
 * Json Serialization
 ****************************************************************/

fun Geofence.toJson(): String {
    return Json(JsonConfiguration.Stable).stringify(Geofence.serializer(), this)
}

/**
 * @throws [kotlinx.serialization.json.JsonDecodingException]
 */
fun String.toGeofence(): Geofence {
    return Json(JsonConfiguration.Stable).parse(Geofence.serializer(), this)
}

object ShapeWKTSerializer : KSerializer<Shape> {

    override val descriptor: SerialDescriptor = StringDescriptor

    override fun serialize(encoder: Encoder, obj: Shape) {
        encoder.encodeString(GEO.formats.wktWriter.toString(obj))
    }

    override fun deserialize(decoder: Decoder): Shape {
        val reader = GEO.formats.wktReader as CustomWKTReader
        return reader.parse(decoder.decodeString()) as Shape
    }
}
